runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/ReferenceCountingFlinkExecutableStageContextFactory.java

```diff
@@ -104,9 +109,30 @@ private void scheduleRelease(JobInfo jobInfo) {
     WrappedContext wrapper = getCache().get(jobInfo.jobId());
     Preconditions.checkState(
         wrapper != null, "Releasing context for unknown job: " + jobInfo.jobId());
-    // Do not release this asynchronously, as the releasing could fail due to the classloader not being
-    // available anymore after the tasks have been removed from the execution engine.
-    release(wrapper);
+
+    PipelineOptions pipelineOptions =
+        PipelineOptionsTranslation.fromProto(jobInfo.pipelineOptions());
+    int environmentCacheTTLMillis =
+        pipelineOptions.as(PortablePipelineOptions.class).getEnvironmentCacheMillis();
+    if (environmentCacheTTLMillis > 0) {
+      if (this.getClass().getClassLoader() != ExecutionEnvironment.class.getClassLoader()) {
```

Comment:
We should add a comment here. This is not self-explanatory.

Reply:
Are we sure that always works? What if we ship ExecutionEnvironment with the child classloader, e.g. in a fat jar? IMHO safer here would be to check for `FlinkUserCodeClassLoader` and see if it is still open.

Reply:
That's what I tested, see above. Adding a comment would be nice, can be done as part of #6592

Reply:
Above you said "not tested" :) Yes, we can add a comment with the other PR.

Reply:
Wordcount was a negative test (with fat jar): Since environment cannot be cached because the class loader isn't the same, it is expected to fail. The test that I did not perform was to add the jar to the Flink classpath (the objective was to optimize embedded execution for VR tests).

Reply:
Added the comment in PR #6592 
>  What if we ship ExecutionEnvironment with the child classloader, e.g. in a fat jar?
Yes, that's a potential bug.

Reply:
That is exactly what we do today.

Reply:
Interesting. In that case, the environment caching will not work.
What is the advantage of loading the ExecutionEnvironment on child classloader?

Reply:
We should clarify in the comment that this in an optimization which only works for embedded execution or when the user code has been added to the Flink classpath (e.g. via the lib folder). Potentially we also need to invert class loading behavior via the Flink config to ensure that the parent is searched first.

Reply:
I think the above code is more explicit with:

```java
 if (!(this.getClass().getClassLoader() instanceof FlinkUserCodeClassloader)) {
```

Reply:
> What is the advantage of loading the ExecutionEnvironment on child classloader?

No advantage, but it is the default behavior of Flink to look in the child classloader first.